### PR TITLE
Toolbar: Add unstyled variant

### LINKED ISSUE #54868: Toolbar: set toolbar variant only when in block toolbar
> Yea, I couldn't identify why the default was applying is-alternate here, so setting undefined variant resulted in the standard shadow.

Since https://github.com/WordPress/gutenberg/pull/51154, changes were made so that `DropdownMenu` and `Dropdown` components rendered inside `Toolbar` would automatically pick the `toolbar` variant — that is achieved through the internal components context system (see [here](https://github.com/WordPress/gutenberg/blob/e5349611de3523240ce169dd7132ecf753c71481/packages/components/src/toolbar/toolbar/index.tsx#L22-L29)) and it causes the popover to render [with the `is-alternate` classname.
](https://github.com/WordPress/gutenberg/blob/397b2bf4fdc57866383deb31d69dd05778c4b505/packages/components/src/popover/index.tsx#L391-L395)
So, given the previous requirements, the popover is currently behaving like expected: it's rendering a black border because it's rendered inside a `Toolbar` component.

From what I understand, the requirements that we gathered initially were incomplete, and what we'd really want is the `toolbar` popover variant to be only applied in popovers opened from the block toolbar (and not from _any_ toolbar).

> Therefore, this issue might need to be resolved in a higher-level component. Otherwise, similar problems may occur when new popovers are added to the header in the future.

That is correct. I guess for now, we could keep the changes from this PR. At the same time, we could work on refining the design requirements once for all, and then apply them at the components level.

_Originally posted by @ciampo in https://github.com/WordPress/gutenberg/issues/54840#issuecomment-1737390574_

### PR DESCRIPTION
**Pull request by brookewp**
## What?

Closes #54868

This covers a missing requirement where the `Toolbar` component shouldn't always use the `toolbar` variant of `Popover` for `DropdownMenu` when in the block editor. @jasmussen [clarified the requirements](https://github.com/WordPress/gutenberg/issues/54868#issuecomment-1744371875): 

>the dropdown menu should be made of the same material as the container that opened it

## Why?

This was discovered as a bug in #54840 and a fix was added to remove the variant. 

As a longer-term solution, @ciampo suggested we add a variant to `Toolbar` so we can remove existing overrides (which also include CSS overrides) and avoid adding more in the future. 

## How?

As suggested, an 'unstyled' `variant` has been added to `Toolbar`. If this is set, the `DropdownMenu`'s `Popover` will be set to the default (`undefined`) value.

This PR also removes the overrides added, the one mentioned above, and CSS overrides. See these changes in the testing steps below.

<details><summary><h2>Testing Instructions</h2></summary>

The main thing to look for is that the `Toolbar` components appear as they did before, with no regressions. It should only appear differently in the blocks where an 'unstyled' `variant` has been set. 

### BlockContextualToolbar

The only change here should be on smaller screens when the toolbar becomes full width and to a lighter border. Its dropdown should now match: 

| Before  | After |
| ------------- | ------------- |
| Screenshot 2023-10-02 at 10 36 58 AM | Screenshot 2023-10-06 at 11 00 58 AM  |

Otherwise, the block toolbar and its dropdowns should appear as it did before: 

Screenshot 2023-10-06 at 11 01 07 AM

### HeaderToolbar

The override added in #54840 is removed from `ToolSelector` but is resolved by adding the variant to `HeaderToolbar`, which allows us to remove the CSS override added before. This should continue to appear the same as it does in trunk currently. 

Screenshot 2023-10-06 at 11 00 47 AM

</details>

## ✍️ Dev note

A new prop, `variant`, has been added to the Toolbar component to add more flexibility in styling. This allows consumers to adopt an `'unstyled'` option which removes default border styles, so fewer manual overrides will be needed. If the prop is `undefined`, the Toolbar will retain its default border styles.

### CODE REVIEW

File: packages/components/src/toolbar/toolbar/types.ts [hunk 1]
```diff
@@ -12,4 +12,14 @@ export type ToolbarProps = {
...
+	/**
+	 * Specifies the toolbar's style.
+	 *
+	 * Leave undefined for the default style. Or 'unstyled' which
+	 * removes the border from the toolbar, but keeps the default
+	 * popover style.
+	 *
+	 * @default undefined
+	 */
+	variant?: 'unstyled' | undefined;
```
Comment (brookewp) [inline packages/components/src/toolbar/toolbar/types.ts:24]: I called this 'unstyled' to match what is used for a borderless style in the `Popover` component. 

@ciampo - you've been mentioning 'borderless' so if you prefer another name, I'm happy to change. I just thought this worked for consistency.

File: packages/components/src/toolbar/stories/index.story.tsx [hunk 2]
```diff
@@ -181,3 +185,42 @@ WithoutGroup.args = {
 		</>
 	),
 };
+
+/**
+ * Set the variant to `unstyled` to remove default border styles.
+ * Otherwise, leave it as `undefined` for default styles.
+ */
+
+export const WithoutStyles = Template.bind( {} );
```
Comment (brookewp) [inline packages/components/src/toolbar/stories/index.story.tsx]: I mentioned consistency with Popover...

So if we decide to keep the name 'unstyled', I'll update the story name to match.

File: packages/components/src/toolbar/toolbar/types.ts (hunk 1 above)
Comment (brookewp) [reply packages/components/src/toolbar/toolbar/types.ts:24]: Along these lines, I also wondered if we'd want `Toolbar` to have a `'popover'` `variant` to match `Popover`'s `undefined` style. But I don't think it would necessarily be that helpful; for example, in the one override addressed in this PR, the entire border is removed, and then just a bottom border is added to replace it: 

https://github.com/WordPress/gutenberg/blob/c6718435c545175675d883d1f5003b5f97a53b8e/packages/block-editor/src/components/block-tools/style.scss#L112-L113

File: packages/components/src/toolbar/toolbar/index.tsx [hunk 3]
```diff
@@ -55,8 +62,10 @@ function UnforwardedToolbar(
 	// `ToolbarGroup` already uses components-toolbar for compatibility reasons.
 	const finalClassName = classnames(
 		'components-accessible-toolbar',
-		className
+		className,
+		variant === undefined ? undefined : `is-${ variant }`
```
Comment (tyxla) [inline packages/components/src/toolbar/toolbar/index.tsx]: `classNames()` is flexible and works just fine with both `undefined` and `false` values, so this shorter version will also work well:
```suggestion
		variant && `is-${ variant }`
```

File: packages/edit-post/src/components/header/header-toolbar/index.js [hunk 4]
```diff
@@ -136,6 +136,7 @@ function HeaderToolbar( { setListViewToggleElement } ) {
 			className="edit-post-header-toolbar"
 			aria-label={ toolbarAriaLabel }
 			shouldUseKeyboardFocusShortcut={ ! blockToolbarCanBeFocused }
+			variant={ 'unstyled' }
```
Comment (tyxla) [inline packages/edit-post/src/components/header/header-toolbar/index.js]: When working with strings, it's more straightforward/readable to just pass it in the string attribute value style:

```suggestion
			variant="unstyled"
```

otherwise, the difference is very minor and boils down to a `JSXExpressionContainer` vs `JSXIdentifier` under the hood, where one can further be evaluated and the other is already evaluated.

File: packages/components/src/toolbar/toolbar/index.tsx [hunk 5]
```diff
@@ -19,23 +19,30 @@ import type { ToolbarProps } from './types';
...
-	Dropdown: {
-		variant: 'toolbar',
-	},
-};
-
 function UnforwardedToolbar(
 	{
 		className,
 		label,
+		variant,
```
Comment (tyxla) [inline packages/components/src/toolbar/toolbar/index.tsx:26]: I think this prop is worth adding to the component README, maybe with some short examples.

File: packages/components/src/toolbar/toolbar/index.tsx [hunk 6]
```diff
@@ -19,23 +19,30 @@ import type { ToolbarProps } from './types';
...
+
+		return {
+			DropdownMenu: {
+				variant: 'toolbar',
+			},
+			Dropdown: {
+				variant: 'toolbar',
+			},
+		};
+	}, [ variant ] );
```
Comment (tyxla) [inline packages/components/src/toolbar/toolbar/index.tsx]: `CONTEXT_SYSTEM_VALUE` will be recalculated on every value, regardless what it is. At the same time, we only care about its declaration if it changes from `undefined` to any other value, or the other way around. 

So, to micro-optimize, I'd recommend declaring that as a value above the `useMemo` and using that instead of `variant`:

```suggestion
	const isVariantDefined = variant !== undefined;
	const CONTEXT_SYSTEM_VALUE = useMemo( () => {
		if ( isVariantDefined ) {
			return {};
		}

		return {
			DropdownMenu: {
				variant: 'toolbar',
			},
			Dropdown: {
				variant: 'toolbar',
			},
		};
	}, [ isVariantDefined ] );
```

File: packages/components/src/toolbar/toolbar/index.tsx [hunk 7]
```diff
@@ -19,23 +19,30 @@ import type { ToolbarProps } from './types';
...
 function UnforwardedToolbar(
 	{
 		className,
 		label,
+		variant,
 		...props
 	}: WordPressComponentProps< ToolbarProps, 'div', false >,
 	ref: ForwardedRef< any >
 ) {
+	const CONTEXT_SYSTEM_VALUE = useMemo( () => {
```
Comment (tyxla) [inline packages/components/src/toolbar/toolbar/index.tsx]: I'd encourage changing `CONTEXT_SYSTEM_VALUE` to `contextSystemValue` since this is no longer a module constant.

**Review by tyxla — COMMENTED** [review]
I'd let the API design review to others who already have more context, but in the meantime, I've left some minor drive-by improvement suggestions.

**Comment by jasmussen** [comment]
Thanks for keeping on this! From just discerning the before and after, this looks right, good job!

I'll defer to others for a code review, but let me know if you need me to dive in and test more deeply.

File: packages/components/src/toolbar/toolbar/index.tsx [hunk 8]
```diff
@@ -19,23 +19,30 @@ import type { ToolbarProps } from './types';
...
+		if ( variant !== undefined ) {
+			return {};
```
Comment (ciampo) [inline packages/components/src/toolbar/toolbar/index.tsx:34]: Probably returning `undefined` is better than an empty object in this case

File: packages/components/src/toolbar/toolbar/types.ts (hunk 1 above)
Comment (ciampo) [reply packages/components/src/toolbar/toolbar/types.ts:24]: Good point about consistency. In general, I like to use names that are more based on the semantics of the variant, rather than the literal aspect that changes. In this case, my previous suggestion of "borderless" is NOT a good name, because it's very literal: what happens if, in the future, we decide to change `Toolbar`'s default styles not to have borders?

I think `unstyled` is a better choice, both because it's a bit more generic (semantic) and because it's more consistent with `Popover`. 

> Along these lines, I also wondered if we'd want Toolbar to have a 'popover' variant to match Popover's undefined style

I don't think it would be much of help at this point in time.


If anything, having the `Popover` exposing a `toolbar` variant now doesn't make much sense, but I guess we can worry about that in a follow-up.

File: packages/components/src/toolbar/test/index.tsx [hunk 9]
```diff
@@ -25,5 +25,13 @@ describe( 'Toolbar', () => {
 				screen.getByLabelText( 'control2', { selector: 'button' } )
 			).toBeInTheDocument();
 		} );
+
+		it( 'should render a toolbar without styles if variant has been defined', () => {
```
Comment (ciampo) [inline packages/components/src/toolbar/test/index.tsx]: Since the "without styles" part is not accurate (it still has some styles), let's maybe rephrase this to something like

```suggestion
		it( 'should apply the unstyled variant correctly via the `variant` prop', () => {
```

File: packages/components/src/toolbar/stories/index.story.tsx (hunk 2 above)
Comment (ciampo) [reply packages/components/src/toolbar/stories/index.story.tsx]: > I'll update the story name to match.

Yeah, "Unstyled" sounds like a better name for this story.

Also, I think we could get away with re-using the default story's arguments — ie.

```tsx
WithoutStyles.args = {
	...Default.args,
	variant: 'unstyled',
};
```

And even more in general, I wonder if we could avoid adding a new story altogether, since folks could just use the `variant` controls on existing stories?

**Review by ciampo — COMMENTED** [review]
This is testing well for me too. Left a couple of suggestions!

**Review by chad1008 — COMMENTED** [review]
Nice PR @brookewp  Looks good to me once the existing feedback from others is addressed!

File: packages/components/src/toolbar/stories/index.story.tsx (hunk 2 above)
Comment (brookewp) [reply packages/components/src/toolbar/stories/index.story.tsx]: We could leave the story out. I added it because it's a visual change, so I thought seeing an explicit demo could be helpful. Also, `Popover` has the same story so I can add another point to my consistency list 😄 

But I'm happy to remove it if you don't think it adds enough value.

File: packages/edit-post/src/components/header/header-toolbar/index.js (hunk 4 above)
Comment (brookewp) [reply packages/edit-post/src/components/header/header-toolbar/index.js]: Good catch! You've shared this with me before, and so I actively try to remember, but this one got away. 🙂

File: packages/components/src/toolbar/toolbar/index.tsx (hunk 3 above)
Comment (brookewp) [reply packages/components/src/toolbar/toolbar/index.tsx]: Thank you! Much cleaner

File: packages/components/src/toolbar/toolbar/index.tsx (hunk 8 above)
Comment (brookewp) [reply packages/components/src/toolbar/toolbar/index.tsx:34]: That was my initial thought, but that resulted in a type error: 

>Type '{ DropdownMenu: { variant: string; }; Dropdown: { variant: string; }; } | undefined' is not assignable to type 'Record<string, any>'.
>Type 'undefined' is not assignable to type 'Record<string, any>'.

Any ideas?

File: packages/components/src/toolbar/toolbar/index.tsx (hunk 6 above)
Comment (brookewp) [reply packages/components/src/toolbar/toolbar/index.tsx]: Thanks for sharing! I'm unfamiliar with `useMemo` and implemented it due to a `console.warn`. So, I'm unsure if I fully understand the difference. 🤔 If you have time to share more or if you know of any examples I could look into to understand further, I'd appreciate it!

File: packages/components/src/toolbar/toolbar/index.tsx (hunk 7 above)
Comment (brookewp) [reply packages/components/src/toolbar/toolbar/index.tsx]: I'm also unfamiliar with the term 'module constants' so thanks for suggesting! My assumption on this change is now there is more than one possible return value vs. before. Is that correct, or is it based on something else?

File: packages/components/src/toolbar/toolbar/index.tsx (hunk 8 above)
Comment (ciampo) [reply packages/components/src/toolbar/toolbar/index.tsx:34]: Ok, it looks like the `ContextSystemProvider` component doesn't accept an undefined `value`. We can keep things as they are :)

File: packages/components/src/toolbar/toolbar/index.tsx (hunk 6 above)
Comment (ciampo) [reply packages/components/src/toolbar/toolbar/index.tsx]: The code inside `useMemo` gets executed only when any of its dependencies change — it is therefore useful when we want to only re-compute something when absolutely necessary, and otherwise use the result from past computations. This is particularly useful for a bunch of reasons:
- it's great when we need to compute something particularly complex / slow to calculate (not this case)
- since React components re-render any time one of their props change, by memoizing a value we can try to reduce that number of rerenders. In fact, without `useMemo` we would be technically creating a new, different object every time

The [react docs](https://react.dev/reference/react/useMemo) are a good start if you're interested in learning more

File: packages/components/src/toolbar/toolbar/index.tsx (hunk 7 above)
Comment (ciampo) [reply packages/components/src/toolbar/toolbar/index.tsx]: This simply a naming convention: constant values variables are usually UPPERCASE_LIKE_THIS. That was the case for `CONTEXT_SYSTEM_VALUE` before this PR.

But since it's been moved inside the react component's function, it's no longer a constant. Marin is suggesting that we change the casing (if that's even a word) to a more traditional camelCase

File: packages/components/src/toolbar/toolbar/index.tsx (hunk 6 above)
Comment (tyxla) [reply packages/components/src/toolbar/toolbar/index.tsx]: What @ciampo said 👍 

In terms of how often it recomputes, `useMemo()` is very similar to `useEffect` or `useCallback` - it will be recomputed only if one of the dependencies change, so we always strive to keep the dependencies as minimal and as static as possible. 

Let's consider an example. If `variant` is a dependency, then the `useMemo()` will be recompute every time `variant` changes, for example, if `variant` was `A` and gets changed to `B`, or if it was `B` and was changed to `undefined`. However, if we declare `const isVariantDefined = variant !== undefined;` and then use `isVariantDefined` as a dependency, `useMemo` will recompute only if `isVariantDefined` changes from `undefined` to anything else, or from anything else to `undefined`. That means, if `variant` changes from `A` to `B`, `isVariantDefined` won't change, therefore `useMemo()` won't recompute.

That being said, it's a micro-optimization, and likely won't have a measurable impact, unless we change the `variant` prop often for some reason.

File: packages/components/src/toolbar/toolbar/index.tsx (hunk 7 above)
Comment (tyxla) [reply packages/components/src/toolbar/toolbar/index.tsx]: Thanks for elaborating, @ciampo, that's exactly what I meant.

**Review by tyxla — APPROVED** [review]
From my perspective, this is good to go 👍 

Feel free to 🚀 once you address the rest of the feedback (@ciampo brought up a suggestion  the Storybook stories)

File: packages/components/src/toolbar/stories/index.story.tsx (hunk 2 above)
Comment (ciampo) [reply packages/components/src/toolbar/stories/index.story.tsx]: Let's keep it for now, but maybe rewrite the `args` to reuse `Default.args` like suggested above?

Comment (brookewp) [reply packages/components/src/toolbar/stories/index.story.tsx]: Sounds good!
